Implement PDF readers

Ts1n opened the pull request:
- Add [tabula](https://github.com/tabulapdf/tabula-java) and [Itext](https://github.com/itext/itext7) dependencies
- Add `PdfExtractor` interface and `PdfExtractorException`
- Add two `PdfExtractor` implementations, `TabulaPdfExtractor` and `ITextPdfExtractor`
- Add respective tests

Closes #35 , closes #36

Ts1n committed:
Add Tabula and Itext dependencies

Ts1n committed:
Add PdfExtractor interface

Ts1n committed:
Add PdfExtractorException

Ts1n committed:
Add TabulaPdfExtractor

Ts1n committed:
Add TabulaPdfExtractorTests

Ts1n committed:
Add ITextPdfExtractor

Ts1n committed:
Add ITextPdfExtractorTests

sscosta commented on src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt:
```diff
@@ -0,0 +1,38 @@
+class ITextPdfExtractor : PdfExtractor {
+    /**
+     * Extract text data from pdf file locate at [pdfPath]
+     * @return Pair<Boolean, String>
```
this might have been left from a previous version

sscosta commented on src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt:
```diff
@@ -0,0 +1,38 @@
+
+        reader.close()
+
+        return Try.of(mutableListOf(data.toString()))
```
as data is MutableList<String>, isn't this the same as "return Try.of(data)"?

sscosta commented on src/main/kotlin/org/ionproject/integration/extractor/implementation/TabulaPdfExtractor.kt:
```diff
@@ -0,0 +1,42 @@
+class TabulaPdfExtractor : PdfExtractor {
+    /**
+     * Extract table data from pdf file locate at [pdfPath]
+     * @return Pair<Boolean, String>
```
same as iText extractor

sscosta commented on src/main/kotlin/org/ionproject/integration/extractor/implementation/TabulaPdfExtractor.kt:
```diff
@@ -0,0 +1,42 @@
+        val parser = DefaultParser()
+        val cmd = parser.parse(CommandLineApp.buildOptions(), args)
+
+        // CommandLineApp(data, cmd).extractTables(cmd)
```
this might have been used in debug

sscosta reviewed (dismissed):
saw some minor adjustments

Ts1n committed:
Code cleanup

Ts1n committed:
Fix incorrect return statement

Ts1n committed:
Wrap IO calls inside withContext using coroutine dispacher for io tasks

sscosta reviewed (dismissed):
Looks great!

pmhsfelix commented on src/main/kotlin/org/ionproject/integration/extractor/interface/PdfExtractor.kt:
```diff
@@ -0,0 +1,13 @@
+     *     Success - String list contains extracted data
+     *     Failure - PdfExtractorException
+     */
+    suspend fun extract(pdfPath: String): Try<MutableList<String>>
```
Why a mutable list?

pmhsfelix replied:
What is this  `pdfPath`? Can't it be a more strongly-typed _thing_ (e.g `URI` or `File`)?

Ts1n replied:
> 
> 
> Why a mutable list?

Data obtained via `iText` comes in string format and it's easier to process the data page by page.

> 
> 
> What is this `pdfPath`? Can't it be a more strongly-typed _thing_ (e.g `URI` or `File`)?

`pdfPath` is the path to the pdf file. It has to be a path because `Tabula` is executed as CLI.

pmhsfelix replied:
From the following sentence present at [https://github.com/tabulapdf/tabula-java/], I got the idea that it is possible to use Tabula as a library
```
You can also integrate tabula-java with any JVM language. For Java examples, see the tests folder.
```

Ts1n replied:
From what I was able to gather we should use `CommandLineApp` and that receives a file path. The alternative would be to use [PDFBox](https://pdfbox.apache.org/) that [tabula uses](https://github.com/tabulapdf/tabula-java/blob/master/pom.xml#L238), but that would negate the benefits of using `tabula`...

pmhsfelix commented on src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt:
```diff
@@ -0,0 +1,42 @@
+        if (pdfPath.isEmpty()) return Try.ofError(PdfExtractorException("Empty path"))
+
+        val tryReader = Try.of { PdfReader(pdfPath) }
+        if (tryReader is Try.Error) return Try.ofError(PdfExtractorException("File doesn't exist or Invalid file format"))
```
We should need to do these explicit `ifs` when using `Try`. Did you face any blocking issues?

Ts1n replied:
`PdfReader(pdfPath)` throws an `java.io.IOException`, but we want all errors to be of type `PdfExtractorException` for easier error handling.

pmhsfelix replied:
We can add a `mapError` to the `Try` to do this mapping.

pmhsfelix commented on src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt:
```diff
@@ -0,0 +1,42 @@
+            data.add((result as Try<String>).toString())
+        }
+
+        reader.close()
```
Don't we need to ensure the `reader` is always closed?

Ts1n replied:
We do. I'll fix it

pmhsfelix commented on src/test/kotlin/org/ionproject/integration/extractor/ITextPdfExtractorTests.kt:
```diff
@@ -0,0 +1,76 @@
+
+        // Assert
+        Assertions.assertEquals(true, result is Try.Value<MutableList<String>>)
+        Assertions.assertNotEquals(emptyList<String>(), (result as Try.Value<MutableList<String>>).value)
```
We need a better way to do this

Ts1n replied:
In what way?

pmhsfelix replied:
Avoid the explicit cast, eventually by creating a helper function with it.

pmhsfelix commented on src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt:
```diff
@@ -0,0 +1,42 @@
+     * Extract text data from pdf file locate at [pdfPath]
+     * @return Try<MutableList<String>>
+     *     Success - String list contains all extracted data in text format
+     *     Failure - PdfExtractorException
```
We shouldn't need to add these as comments. It should be explicit on the types. Perhaps we need to encode the exception type on the `Try`.

pmhsfelix replied:
The returned type should document what are the returned options. Unfortunately, the `Try` design allows that for the success case but not for the error case.

pmhsfelix commented on src/main/kotlin/org/ionproject/integration/extractor/implementation/TabulaPdfExtractor.kt:
```diff
@@ -0,0 +1,44 @@
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.apache.commons.cli.DefaultParser
+import org.ionproject.integration.extractor.`interface`.PdfExtractor
```
Use plural (`interfaces`) to avoid using a reserved keyword.

pmhsfelix commented on src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt:
```diff
@@ -0,0 +1,42 @@
+
+        val data = mutableListOf<String>()
+        for (i in 1..pdfDoc.numberOfPages) {
+            val result = withContext(Dispatchers.IO) {
```
Why is this needed?
What are we trying to achieve with this `withContext`, since `PdfTextExtractor.getTextFromPage` is not suspend?

pmhsfelix replied:
From the `withContext` docs:
```
Calls the specified suspending block with a given coroutine context, suspends until it completes, and returns the result.
```
So, it seems we are just changing the thread that gets blocked. Namely, the iteration will not proceed while the `getTextFromPage` doesn't complete.

pmhsfelix commented on src/main/kotlin/org/ionproject/integration/extractor/implementation/TabulaPdfExtractor.kt:
```diff
@@ -0,0 +1,44 @@
+        val cmd = parser.parse(CommandLineApp.buildOptions(), args)
+
+        val data = StringBuilder()
+        var result = withContext(Dispatchers.IO) {
```
Same

pmhsfelix commented on src/main/kotlin/org/ionproject/integration/extractor/implementation/ITextPdfExtractor.kt:
```diff
@@ -0,0 +1,42 @@
+import org.ionproject.integration.extractor.exception.PdfExtractorException
+import org.ionproject.integration.utils.Try
+
+class ITextPdfExtractor : PdfExtractor {
```
Why does it start with an `I`?

Ts1n replied:
The tool is named [`IText`](https://github.com/itext/itext7)

Ts1n committed:
Merge branch 'master' into feature/35_36

Ts1n committed:
Rename folders

Ts1n committed:
Remove suspend modifier

Given that there is not UI thread, there is no need for async work. If needed, we'll readd in the future

Ts1n committed:
Refactor PdfExtractor Implementations

Correct use of Try

Ts1n committed:
Refactor PdfExtractor tests

Ts1n committed:
Refactor Try test

Ts1n committed:
Moved tests for package consistency

sscosta commented on src/main/kotlin/org/ionproject/integration/extractor/implementations/ITextPdfExtractor.kt:
```diff
@@ -0,0 +1,36 @@
+     * Extract text data from pdf file locate at [pdfPath]
+     * @return CompositeException with list of exceptions in case of any error
+     */
+    override fun extract(pdfPath: String): Try<MutableList<String>> {
```
I think we should return Try<List< String>>

sscosta commented on src/main/kotlin/org/ionproject/integration/extractor/implementations/ITextPdfExtractor.kt:
```diff
@@ -0,0 +1,36 @@
+class ITextPdfExtractor : PdfExtractor {
+    /**
+     * Extract text data from pdf file locate at [pdfPath]
+     * @return CompositeException with list of exceptions in case of any error
```
It also returns PDFExtractorException when path is empty

sscosta commented on src/main/kotlin/org/ionproject/integration/extractor/implementations/ITextPdfExtractor.kt:
```diff
@@ -0,0 +1,36 @@
+                .mapError { PdfExtractorException("Itext cannot process file") }
+        } finally {
+            pdfReader
+                .match({ reader -> reader.close() }, {})
```
As we're not using the second function of match, i think it is equivalent to using the map function

sscosta commented on build.gradle.kts:
```diff
@@ -24,6 +24,11 @@ dependencies {
+        exclude(group = "org.slf4j")
+    }
+    implementation("com.itextpdf:kernel:7.0.0")
+    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.3.5")
```
For now we will not need coroutines

Ts1n committed:
Remove coroutine dependency

Ts1n committed:
Replace .match with .map

Ts1n committed:
Refactor comments

Ts1n committed:
Format

Ts1n committed:
Refactor extract to return List<String>
